# ADAPTIVEMASKING

Examine NGS mapped output to identify regions of high minor variant frequencies, as [described]( 
https://www.biorxiv.org/content/early/2018/01/23/252460).

## DOCUMENT IN PROGRESS: DATA STILL BEING ADDED
Here, we describe software to conduct the analyses we describe in our paper.

__Obtain software and test data__   
The software needed, and how to obtain test data illustrating end-to-end processing, as well as the inputs into the modelling described in our paper, are [here](doc/Prerequisites.md).  

__Overview of the process followed__  
1. Mapping and VCF generation  
To aid testing of our pipeline we describe end-to-end analysis including an [illustration of mapping and vcf generation](doc/map2vcf.md) operations on a test dataset.
However, it is important to understand that our approach is designed to operate on the output from multiple mappers, with different settings, and different kinds of input data.
The objective of the project is quantify the variation associated with the combination of laboratory process, input DNA, sequencing, and mapping.
Many pipelines will have already optimised mapping tools and settings.  The process we describe will operate on such output.
It expects VCF files, which are typically generated by samtools/bcftools mpileup commands following mapping.  In particular, it expects a tag in the VCF INFO section which contains the high quality base counts at each position: it is these which are the input to the algorithm.
For more details of this requirement are described [below](doc/extractmaf.md).

2. Estimating the amount of extraneous bacterial DNA present
One of the key findings from our paper is that mapping accuracy for some regions is determined by the amount of 'non-target' bacteria DNA present, in our case from species other than *Mycobacteria*.
We estimated this [using Kraken](doc/kraken.md).

3. Determining the minor variant frequencies for genomic regions
[We measured the minor variant frequencies](doc/extractmaf.md) - that is, the read depth accounted for by calls other than the most common nucleotide - across genomic regions.

4. Modelling minor variant frequencies in reads mapped to genomic regions 
This is described here: [Modelling minor variant frequencies](doc/model_maf.md)
*it isn't at present*

5. Depicting the results of the modelling
See here: [Depicting the results](doc/depict.md)

6. Masking based on the results
[What to do next](doc/next.md) 

