# ADAPTIVEMASKING

We describe tools to examine NGS-mapped output to identify regions of high minor variant frequencies, as [described](https://www.biorxiv.org/content/early/2018/01/23/252460).  The rationale of doing this is the identification of regions where consensus basecalling may be unreliable.  The process is 'adaptive' in the sense that it models variation which arises during the entire laboratory, sequencing and mapping process and can be used to identify regions of variation resulting anywhere in the process followed.  In the situation [described](https://www.biorxiv.org/content/early/2018/01/23/252460), increased variation occurred in regions of homology between Mycobacteria and other bacterial species, and was detected when laboratory processes started to use broth-culture derived DNA extracts, rather than extracts from pure cultures.

__Obtain software and test data__   
The software needed, and how to obtain test data illustrating end-to-end processing, as well as the inputs into the modelling described in our paper, are [here](doc/Prerequisites.md).  

__Overview of the process followed__  
1. Mapping and VCF generation  
To aid testing of our pipeline we describe end-to-end analysis including an [illustration of mapping and vcf generation](doc/map2vcf.md) operations on a test dataset.
However, it is important to understand that our approach is designed to operate on the output from multiple mappers, with different settings, and with different kinds of input data.
The objective of the project is quantify the variation associated with the totality of laboratory process, input DNA, sequencing, and mapping.
Many pipelines will have already optimised mapping tools and settings; the process we describe will operate on their output.  As input it expects VCF/BCF files, which are typically generated by samtools/bcftools mpileup commands following mapping.  In particular, it expects a tag in the VCF INFO section which contains the high quality base counts at each position: it is these which are the input to the algorithm.
More details of this requirement are described [below](doc/extractmaf.md).

2. Estimating the amount of extraneous bacterial DNA present    
One of the key findings from our paper is that mapping accuracy for some regions is determined by the amount of 'non-target' bacteria DNA present, in our case from species other than *Mycobacteria*.
We estimated this [using Kraken](doc/kraken.md).  Newer [tools](https://ccb.jhu.edu/software/bracken/) producing data in a similar format can also be used. 

3. Determining the minor variant frequencies for genomic regions  
[We measured the minor variant frequencies](doc/extractmaf.md) - that is, the read depth accounted for by calls other than the most common nucleotide - across genomic regions.

4. Modelling minor variant frequencies in reads mapped to genomic regions     
This is described here: [Modelling minor variant frequencies](doc/model_maf.md)
*it isn't at present*

5. Depicting the results of the modelling
See here: [Depicting the results](doc/depict.md)

6. Masking based on the results
[What to do next](doc/next.md) 

